python/lsst/pipe/tasks/setPrimaryFlags.py

```diff
@@ -69,47 +75,58 @@ def run(self, sources, skyMap, tractInfo, patchInfo, includeDeblend=True):
 
-        tractId = tractInfo.getId()
-        for source in sources:
-            centroidPos = source.getCentroid()
-            if numpy.any(numpy.isnan(centroidPos)):
-                continue
-            if source.getCentroidFlag():
-                # Use a slightly smaller box to guard against bad centroids (see above)
-                isPatchInner = shrunkInnerFloatBBox.contains(centroidPos)
-            else:
-                isPatchInner = innerFloatBBox.contains(centroidPos)
-            source.setFlag(self.isPatchInnerKey, isPatchInner)
+            tractId = tractInfo.getId()
+            for source in sources:
+                centroidPos = source.getCentroid()
+                if numpy.any(numpy.isnan(centroidPos)):
+                    continue
+                if source.getCentroidFlag():
+                    # Use a slightly smaller box to guard against bad centroids (see above)
+                    isPatchInner = shrunkInnerFloatBBox.contains(centroidPos)
+                else:
+                    isPatchInner = innerFloatBBox.contains(centroidPos)
+                source.setFlag(self.isPatchInnerKey, isPatchInner)
 
-            skyPos = source.getCoord()
-            sourceInnerTractId = skyMap.findTract(skyPos).getId()
-            isTractInner = sourceInnerTractId == tractId
-            source.setFlag(self.isTractInnerKey, isTractInner)
+                skyPos = source.getCoord()
+                sourceInnerTractId = skyMap.findTract(skyPos).getId()
+                isTractInner = sourceInnerTractId == tractId
+                source.setFlag(self.isTractInnerKey, isTractInner)
 
-            if nChildKey is None or source.get(nChildKey) == 0:
-                for pseudoFilterKey in pseudoFilterKeys:
-                    if source.get(pseudoFilterKey):
-                        isPseudo = True
-                        break
-                else:
-                    isPseudo = False
+                if nChildKey is None or source.get(nChildKey) == 0:
+                    for pseudoFilterKey in pseudoFilterKeys:
+                        if source.get(pseudoFilterKey):
+                            isPseudo = True
+                            break
+                    else:
+                        isPseudo = False
+
+                    source.setFlag(self.isPrimaryKey, isPatchInner and isTractInner and not isPseudo)
 
-                source.setFlag(self.isPrimaryKey, isPatchInner and isTractInner and not isPseudo)
+        else:
+            hasSkySources = True if "sky_source" in sources.schema else False
+            for source in sources:
+                if nChildKey is None or source.get(nChildKey) == 0:
+                    if hasSkySources:
+                        if not source["sky_source"]:
+                            source.setFlag(self.isPrimaryKey, True)
+                    else:
+                        source.setFlag(self.isPrimaryKey, True)
```

Comment:
I'm quite surprised this is so short when the corresponding code for coadds is so long. Please consider pulling out this core bit into its own "actually set primary key now that we've dealt with whatever-TF coadd pseudofilters are" function so there is less code duplication.

Reply:
Are you suggesting that `setFlag` from both the coadd case and the single-frame case should be pulled out of the above if statement and put into its own section, or just that this single-frame case section should be cleaned up? If the latter, we did discuss this at length, but couldn't come up with a clean way to do it that was in keeping with the coadd case above - suggestions welcome!

Reply:
I thought that both cases had code identical to 
```
for source in sources:
     if nChildKey is None or source.get(nChildKey) == 0:
         if hasSkySources:
             if not source["sky_source"]:
                 source.setFlag(self.isPrimaryKey, True)
         else:
             source.setFlag(self.isPrimaryKey, True)
```
but now I see there's a sneaky other `pseudoFilterKeys` loop hiding in the coadd case. Honestly the whole thing feels a bit too meandery in the if/else logic train; I don't want you to spend forever optimizing the coadd side of things when you were just going in trying to add the single frame side of things. But if you can come up with a way to create a "setPrimaryFlag" function that works for both, that would be a lot nicer.

Reply:
So, after having thought about it for a while, I think the only option to achieve this would be to move the for loop one level higher and check whether or not we're working on coadd data or single-frame data. This might make the code clearer, but it would introduce an if-statement check for each iteration of the for loop, which may impact efficiency somewhat. With that said, and keeping the coadd code and the single-frame code separate, I've modified the single frame block to this (below) which should hopefully be more readable? Thoughts welcome!

        else:
            hasSkySources = True if "sky_source" in sources.schema else False
            for source in sources:
                hasNoChildren = True if nChildKey is None or source.get(nChildKey) == 0 else False
                isSkySource = False
                if hasSkySources:
                    if source["sky_source"]:
                        isSkySource = True
                source.setFlag(self.isPrimaryKey, hasNoChildren and not isSkySource)

Reply:
I think this is good - I like `hasNoChildren`, that adds a lot of clarity. It's fine to keep the coadd block and single frame blocks separate to avoid slowing down the looping.